impeller/entity/contents/tiled_texture_contents.cc

```diff
@@ -153,34 +158,64 @@ bool TiledTextureContents::Render(const ContentContext& renderer,
     options.stencil_operation = StencilOperation::kIncrementClamp;
   }
   options.primitive_type = geometry_result.type;
+
+#ifdef IMPELLER_ENABLE_OPENGLES
+  if (is_external_texture) {
+    cmd.pipeline = renderer.GetTiledTextureExternalPipeline(options);
+  } else {
+    cmd.pipeline = uses_emulated_tile_mode
+                       ? renderer.GetTiledTexturePipeline(options)
+                       : renderer.GetTexturePipeline(options);
+  }
+#else
   cmd.pipeline = uses_emulated_tile_mode
                      ? renderer.GetTiledTexturePipeline(options)
                      : renderer.GetTexturePipeline(options);
+#endif  // IMPELLER_ENABLE_OPENGLES
 
   cmd.BindVertices(geometry_result.vertex_buffer);
   VS::BindFrameInfo(cmd, host_buffer.EmplaceUniform(frame_info));
 
-  if (uses_emulated_tile_mode) {
+  if (is_external_texture) {
+    FSExternal::FragInfo frag_info;
+    frag_info.x_tile_mode = static_cast<Scalar>(x_tile_mode_);
+    frag_info.y_tile_mode = static_cast<Scalar>(y_tile_mode_);
+    FSExternal::BindFragInfo(cmd, host_buffer.EmplaceUniform(frag_info));
+  } else if (uses_emulated_tile_mode) {
     FS::FragInfo frag_info;
     frag_info.x_tile_mode = static_cast<Scalar>(x_tile_mode_);
     frag_info.y_tile_mode = static_cast<Scalar>(y_tile_mode_);
     FS::BindFragInfo(cmd, host_buffer.EmplaceUniform(frag_info));
   }
 
-  if (color_filter_) {
-    auto filtered_texture = CreateFilterTexture(renderer);
-    if (!filtered_texture) {
-      return false;
-    }
-    FS::BindTextureSampler(
-        cmd, filtered_texture,
-        renderer.GetContext()->GetSamplerLibrary()->GetSampler(
-            CreateDescriptor(renderer.GetDeviceCapabilities())));
-  } else {
-    FS::BindTextureSampler(
+  if (is_external_texture) {
+    SamplerDescriptor sampler_desc;
+    // OES_EGL_image_external states that only CLAMP_TO_EDGE is valid, so we
+    // emulate all other tile modes here by remapping the texture coordinates.
+    sampler_desc.width_address_mode = SamplerAddressMode::kClampToEdge;
+    sampler_desc.height_address_mode = SamplerAddressMode::kClampToEdge;
+
+    // Also, external textures cannot be bound to color filters, so ignore this
+    // case for now.
+    FSExternal::BindSAMPLEREXTERNALOESTextureSampler(
```

Comment:
I think we cannot actually hit this case with the API we have today (which doesn't allow adding color/image filters directly to external textures) - so perhaps a DCHECK is appropriate.

Reply:
> which doesn't allow adding color/image filters directly to external textures

Where is this disallowed? 🤔 There is no separate API for external textures in Aiks.

Reply:
The optimization only applies to color filters that are on the paint object used in the draw call. External texture API does not accept a paint object so all filters need to be applied by a separate composited later

Reply:
Oh I see, you mean how we expose it to plugins/framework. Added a dcheck. We can easily add support later for other customers if requested.